### PR TITLE
`storage`: refactoring to use `hashicorp/go-azure-sdk`

### LINKED ISSUE #12627: azurerm_storage_account with network_rules doesn't give the error if missing serviceEndpoint but a 404 not found
### Terraform (and AzureRM Provider) Version

```
Terraform v1.0.2
on linux_amd64
+ provider registry.terraform.io/aztfmod/azurecaf v1.2.5
+ provider registry.terraform.io/hashicorp/azurerm v2.68.0
+ provider registry.terraform.io/hashicorp/external v2.1.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
+ provider registry.terraform.io/hashicorp/template v2.2.0
+ provider registry.terraform.io/hashicorp/vault v2.21.0
+ provider registry.terraform.io/terraform-providers/mysql v1.9.0
+ provider registry.terraform.io/winebarrel/mysql v1.10.4
```

### Affected Resource(s)

* `azurerm_storage_account`

### Terraform Configuration Files

```hcl
resource "azurerm_storage_account" "assets" {
  account_replication_type = "GRS"
  account_kind             = "StorageV2"
  account_tier             = "Standard"
  location                 = module.azure-region.location
  name                     = local.storage_account_name
  resource_group_name      = module.rg.resource_group_name

  network_rules {
    default_action = "Deny"
    bypass         = ["Logging", "Metrics", "AzureServices"]
    ip_rules       = [for value in concat(local.admin_cidrs_list, local.customer_cidrs_list) : replace(value, "/(.*)\\/32/", "$1")]

    virtual_network_subnet_ids = [
      data.terraform_remote_state.xxx.outputs.fo_subnet_id
      data.terraform_remote_state.xxx.outputs.bo_subnet_id
    ]
  }
}
```

### Expected Behaviour
If a subnet doesn't have a serviceendpoint should have this error
̀```
│ Error: updating Azure Storage Account network_rules "zhstdeepr4qeuw1": storage.AccountsClient#Update: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="NetworkAclsValidationFailure" Message="Validation of network acls failure: SubnetsHaveNoServiceEndpointsConfigured:Subnets zhsnetdeepadmpeuw1 of virtual network /subscriptions/xxxxxx/resourceGroups/xxxxxx/providers/Microsoft.Network/virtualNetworks/xxxxxxx do not have ServiceEndpoints for Microsoft.Storage resources configured. Add Microsoft.Storage to subnet's ServiceEndpoints collection before trying to ACL Microsoft.Storage resources to these subnets.."
̀```

### Actual Behaviour

An error that doesn't help to debug
```
Error: Error retrieving Azure Storage Account "zhstdeepr4qeuw1": storage.AccountsClient#GetProperties: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="StorageAccountNotFound" Message="The storage account zhstdeepr4qeuw1 was not found."
```

### Steps to Reproduce

1. `terraform apply`

Best regards,

### LINKED ISSUE #14737: [Azure][FileShare][Terraform version = "=2.90.0"] HTTP headers is not in the correct format
│ Error: creating Share "fileshare" (Account "stxsoarconfigslab" / Resource Group "rg-XXXXX-lab"): shares.Client#Create: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="InvalidHeaderValue" Message="The value for one of the HTTP headers is not in the correct format.\nRequestId:61b909b9-501a-0055-3d40-fb496d000000\nTime:2021-12-27T16:39:41.3366258Z"
│ 

image

```
resource "azurerm_storage_account" "storageAccount" {
  name                     = "${var.storage_account_name}"
  resource_group_name      = "${var.resource_group_name}"
  location                 = "${var.location}"
  allow_blob_public_access = true
  account_tier             = "Premium"
  account_kind             = "FileStorage"
  account_replication_type = "LRS"
  tags                     = var.tags
}

resource "azurerm_storage_share" "fileShare" {
  name                 = "${var.file_share_name}"
  storage_account_name = azurerm_storage_account.storageAccount.name
  quota                = "${var.file_share_size}"
  enabled_protocol     = "NFS"

  acl {
    id = "MTIzNDU2Nzg5MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTI"

    access_policy {
      permissions = "${var.file_share_permissoins}"
    }
  }
}
```

### LINKED ISSUE #15609: Fails to create azure container right after azure storage account was created
Getting **The specified resource does not exist** error while creating azure container right after azure storage account was created

### **Source Code :**

```
resource "azurerm_storage_account" "storage_account" {
  name                                = "mystgacnt45612"
  resource_group_name     = "tf-azure"
  location                            = "East US"
  account_tier                     = "Standard"
  account_replication_type = "LRS"
  allow_blob_public_access = true
}

resource "azurerm_storage_container" "storage_container" {
  storage_account_name  = azurerm_storage_account.storage_account.name
  name                              = "mystgcntr45612"
  container_access_type    = "blob"
}
```

### **Actual Behaviour :**
While executing the above script, getting below error
```
│ Error: failed creating container: failed creating container: containers.Client#Create: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="ResourceNotFound" Message="The specified resource does not exist.\nRequestId:652e8c77-701e-008f-024c-2a1e0c000000\nTime:2022-02-25T13:31:39.3693903Z"
│
│   with azurerm_storage_container.storage_container,
│   on main.tf line 10, in resource "azurerm_storage_container" "storage_container":
│   10: resource "azurerm_storage_container" "storage_container" {
│
```

### **Expected Behaviour :**
It should create the both storage account and storage container.

### LINKED ISSUE #18897: Azure Storage Account Provider produced inconsistent result after apply
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Greater than 0.13

### AzureRM Provider Version

2.86.0

### Affected Resource(s)/Data Source(s)

azurerm_storage_account

### Terraform Configuration Files

```hcl
resource "azurerm_storage_account" "main" {
  name                      = "${var.storageaccountname}"
  resource_group_name       = azurerm_resource_group.main.name
  location                  = azurerm_resource_group.main.location
  account_kind              = "StorageV2"
  account_tier              = "Standard"
  enable_https_traffic_only = true
  is_hns_enabled            = true
  account_replication_type  = "LRS"
  access_tier               = "Hot"

  network_rules {
    default_action = "Deny"
    bypass         = ["AzureServices"]
    ip_rules       = var.optum_ip_addresses
  }

  min_tls_version = "TLS1_2"

  tags = var.tags
}
```

### Debug Output/Panic Output

```shell
Error: Provider produced inconsistent result after apply

When applying changes to azurerm_storage_account.main, provider "provider[\"registry.terraform.io/hashicorp/azurerm\"]" produced an unexpected new value: Root resource    
was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own issue tracker.
```

### Expected Behaviour

Storage account should be created

### Actual Behaviour

Error

### Steps to Reproduce

terraform apply

### Important Factoids

_No response_

### References

_No response_

### LINKED ISSUE #19055: Updating the container_access_type property of azurerm_storage_container returns 404
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.3.2

### AzureRM Provider Version

3.29.1

### Affected Resource(s)/Data Source(s)

azurerm_storage_container

### Terraform Configuration Files

```hcl
resource "azurerm_storage_account" "this" {
  name                              = "samplest001"
  resource_group_name               = data.azurerm_resource_group.resource_group.name
  location                          = data.azurerm_resource_group.resource_group.location
  account_tier                      = "Standard"
  account_replication_type          = "LRS"
  access_tier                       = "Hot"
  enable_https_traffic_only         = true
  min_tls_version                   = "TLS1_2"
  shared_access_key_enabled         = false

  network_rules {
    default_action = "Allow"
    bypass         = ["None"]
  }

  identity {
    type = "SystemAssigned"
  }
}

resource "azurerm_storage_container" "this" {
  name                  = "container"
  storage_account_name  = azurerm_storage_account.this.name
  container_access_type = "container"
}
```

### Debug Output/Panic Output

```shell
Error: updating the Access Control for Container "container" (Storage Account "samplest001" / Resource Group "MigrationTest"): containers.Client#SetAccessControl: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="ResourceNotFound" Message="The specified resource does not exist.\nRequestId:36589c2c-e01e-001a-5617-eb5a69000000\nTime:2022-10-28T21:48:39.2004013Z"
```

### Expected Behaviour

Updating the value of container_access_type should perform the same change on the resource in Azure.

### Actual Behaviour

Trying to change this value and running an apply returns a 404 resource not found error.

### Steps to Reproduce

1. terraform init -upgrade -reconfigure -backend-config="backend.cfg"
2. terraform plan -out plan.tfplan
3. terraform apply "plan.tfplan"

### Important Factoids

I was able to reproduce this issue with containers in public storage accounts and storage accounts with private endpoints.

### References

_No response_

### LINKED ISSUE #19427: Terraform tries to update customer_managed_key.key_vault_key_id even if it is in the state
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.0.11

### AzureRM Provider Version

3.32.0

### Affected Resource(s)/Data Source(s)

azurerm_storage_account, azurerm_storage_account_customer_managed_key

### Terraform Configuration Files

```hcl
# Json config 
[
  {
    "name": "hrktestforenc",
    "enabled": true,
    "rgName": "<redacted>",
    "location": "westeurope",
    "accountTier": "Standard",
    "replicationType": "LRS",
    "minTlsVersion": "TLS1_2",
    "hnsEnabled": false,
    "httpsOnly": true,
    "allowPublicAccess": true,
    "accessTier": "Hot",
    "identity": {
      "type": "SystemAssigned",
      "ids": []
    },
    "userManagedEncryption": {
      "enabled": true,
      "vaultName": "<redacted>",
      "rgName": "<redacted>",
      "keyType": "RSA",
      "keySize": "2048"
    },
    "networkRules": {
      "defaultAction": "Allow",
      "bypass": ["AzureServices"],
      "ipRules": [],
      "resourceRules": []
    },
    "queues": [
      {
        "name": "testqueue",
        "meta": {
          "key": "value",
          "param": "othervalue"
        }
      },
      {
        "name": "test2queue",
        "meta": {}
      }
    ]
  }
]

#locals.tf
locals {
  config = jsondecode(file("${path.module}/configs/${var.saenv}/variables.json"))

  encryption = flatten([
    for saccount in local.config : [{
      saccountName = saccount.name
      enabled      = saccount.userManagedEncryption.enabled
      vaultName    = saccount.userManagedEncryption.vaultName
      rgName       = saccount.userManagedEncryption.rgName
      keyType      = saccount.userManagedEncryption.keyType
      keySize      = saccount.userManagedEncryption.keySize
    }]
  ])

  saqueues = flatten([
    for saccount in local.config : [
      for queue in saccount.queues : {
        saccountName = saccount.name
        queueName    = queue.name
        queueMeta    = queue.meta
      }
    ]
  ])

}

#data.tf 
// get the azure vault objects for reference their attributes later
data "azurerm_key_vault" "keyvaults" {
  for_each            = { for encryption in local.encryption : encryption.saccountName => encryption if encryption.enabled }
  name                = each.value.vaultName
  resource_group_name = each.value.rgName
}


## main.tf
# Create storage account
resource "azurerm_storage_account" "accounts" {
  for_each                 = { for saccount in local.config : saccount.name => saccount if saccount.enabled }
  name                     = each.value.name
  resource_group_name      = each.value.rgName
  location                 = each.value.location
  account_tier             = each.value.accountTier
  account_replication_type = each.value.replicationType

  access_tier                   = each.value.accessTier
  enable_https_traffic_only     = each.value.httpsOnly
  min_tls_version               = each.value.minTlsVersion
  public_network_access_enabled = each.value.allowPublicAccess
  is_hns_enabled                = each.value.hnsEnabled

  dynamic "identity" {
    for_each = each.value.identity != null ? toset([each.value.identity]) : toset([])
    content {
      type         = identity.value["type"]
      identity_ids = identity.value["ids"]
    }
  }

  tags = merge(
    {
      Name = each.value.name
    },
    var.ctags
  )




# Add user managed key vault key
resource "azurerm_storage_account_customer_managed_key" "saccountkeys" {
  for_each           = { for encryption in local.encryption : encryption.saccountName => encryption if encryption.enabled }
  storage_account_id = azurerm_storage_account.accounts["${each.value.saccountName}"].id
  key_vault_id       = data.azurerm_key_vault.keyvaults["${each.value.saccountName}"].id
  key_name           = azurerm_key_vault_key.enckey["${each.value.saccountName}"].name
  key_version        = azurerm_key_vault_key.enckey["${each.value.saccountName}"].version

  depends_on = [
    azurerm_storage_account.accounts,
    azurerm_key_vault_key.enckey
  ]
}

# Allow the storage account created systemid to manage the keys in the vault
resource "azurerm_key_vault_access_policy" "storage" {
  for_each     = { for encryption in local.encryption : encryption.saccountName => encryption if encryption.enabled }
  key_vault_id = data.azurerm_key_vault.keyvaults["${each.value.saccountName}"].id
  tenant_id    = azurerm_storage_account.accounts["${each.value.saccountName}"].identity.0.tenant_id
  object_id    = azurerm_storage_account.accounts["${each.value.saccountName}"].identity.0.principal_id

  key_permissions    = ["Get", "Create", "List", "Restore", "Recover", "UnwrapKey", "WrapKey", "Purge", "Encrypt", "Decrypt", "Sign", "Verify"]
  secret_permissions = ["Get"]
}

# Add user managed key vault key
resource "azurerm_storage_account_customer_managed_key" "saccountkeys" {
  for_each           = { for encryption in local.encryption : encryption.saccountName => encryption if encryption.enabled }
  storage_account_id = azurerm_storage_account.accounts["${each.value.saccountName}"].id
  key_vault_id       = data.azurerm_key_vault.keyvaults["${each.value.saccountName}"].id
  key_name           = azurerm_key_vault_key.enckey["${each.value.saccountName}"].name
  key_version        = azurerm_key_vault_key.enckey["${each.value.saccountName}"].version

  depends_on = [
    azurerm_storage_account.accounts,
    azurerm_key_vault_key.enckey
  ]
}
```

### Debug Output/Panic Output

```shell
After plan

# module.storageaccount.azurerm_storage_account.accounts["hrktestforenc"] will be updated in-place
  ~ resource "azurerm_storage_account" "accounts" {
        id                                = "/subscriptions/<redacted>/providers/Microsoft.Storage/storageAccounts/hrktestforenc"
        name                              = "hrktestforenc"
        tags                              = {
            "Name"        = "hrktestforenc"
            "Provisioner" = "aaa"
            "Team"        = "xxx"
        }
        # (36 unchanged attributes hidden)


      - customer_managed_key {
          - key_vault_key_id = "https://<redacted>.vault.azure.net/keys/hrktestforenc-encryption-key/04e0cb9245e1482db93c167a398194a7" -> null
        }




        # (5 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.


# In the state
....
 blob_properties {
        change_feed_enabled           = false
        change_feed_retention_in_days = 0
        last_access_time_enabled      = false
        versioning_enabled            = false
    }

    customer_managed_key {
        key_vault_key_id = "https://<redacted>.vault.azure.net/keys/hrktestforenc-encryption-key/04e0cb9245e1482db93c167a398194a7"
    }

    identity {
        identity_ids = []
...
```

### Expected Behaviour

The state check should recognize the azurerm_storage_account_customer_managed_key resource state as a valid configuration for the azurerm_storage_account resource customer_managed_key section

### Actual Behaviour

During every run it tries to set up the customer_managed_key to null on the azurerm_storage_account regardless that we set it up using azurerm_storage_account_customer_managed_key module

### Steps to Reproduce

terraform apply twice

### Important Factoids

_No response_

### References

_No response_

### LINKED ISSUE #22507: azurerm_storage_account.go non-compliant resource issue
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

1.5.2

### AzureRM Provider Version

3.0.0

### Affected Resource(s)/Data Source(s)

azurerm_storage_account

### Terraform Configuration Files

```hcl
terraform {
  required_version = ">=0.12"
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "=3.0.0"
    }
  }
  backend "azurerm" {}
}
provider "azurerm" {
  features {}
}
```

### Debug Output/Panic Output

```shell
Error: updating Azure Storage Account AllowSharedKeyAccess "***": storage.AccountsClient#Update: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="RequestDisallowedByPolicy" Message="Resource '***' was disallowed by policy.
```

### Expected Behaviour

- Create storage account with no tags
- Create subscription policy to require tags of a specific value
- Using azurerm_storage_account update the tag to the required specified value of the policy
- Tags should be updated and terraform should run alright

### Actual Behaviour

- Create storage account with no tags
- Create subscription policy to require tags of a specific value
- Using azurerm_storage_account update the tag to the required specified value of the policy
- Terraform fails with the policy being raised

Unfortunately can't share images of the code, but in azurerm_storage_account_resource.go on line 1502, I believe that because this client is called regardless to make updates to the storage account, before the client is called to update the tags, the policy is not met and so an error is raised - which essentially softlocks you from making any changes via terraform without manual intervention on the azure portal.

### Steps to Reproduce

1. Using terraform create a storage account with no tags
2. In azure subscription policies, create a policy that requires a particular resource tag
3. The storage account is now non-compliant
4. terraform import that resource into a new script
5. Adjust the resource block to include the tag that the policy requires
6. terraform plan to show clearly that the tag would be added on the run of the terraform
7. terraform apply

This will error because the tag isn't compliant, but I believe it's the blank tag that's not compliant being executed by the client in azurerm_storage_account_resource.go line 1502, and not the actual tags that are updated and should be executed by the client on line 1551

### Important Factoids

N/A

### References

_No response_

### LINKED ISSUE #22992: Error: reading blob properties for AzureRM Storage Account
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

1.5.1

### AzureRM Provider Version

3.56.0

### Affected Resource(s)/Data Source(s)

azurerm_storage_account

### Terraform Configuration Files

```hcl
resource "azurerm_storage_account" "manager" {

  name                = "manager"
  resource_group_name = "myrg"
  location            = "westeurope"

  account_tier             = "Standard"
  account_kind             = "StorageV2"
  account_replication_type = "LRS"
  access_tier              = "Hot"

  enable_https_traffic_only       = true
  shared_access_key_enabled       = true
  allow_nested_items_to_be_public = false

  lifecycle {
    ignore_changes = [
      tags,
    ]
  }
}
```

### Debug Output/Panic Output

```shell
╷
│ Error: reading blob properties for AzureRM Storage Account "stavdmanagerweudev001": storage.BlobServicesClient#GetServiceProperties: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="ContainerOperationFailure" Message="The specified resource does not exist.\nRequestId:781c4f59-301e-0002-7cbe-becdbe000000\nTime:2023-07-25T06:10:54.4910078Z"
│ 
│   with module.avd_manager.azurerm_storage_account.avd_manager,
│   on ../../modules/terraform-azurerm-avd-manager/resources.avd-manager.tf line 132, in resource "azurerm_storage_account" "avd_manager":
│  132: resource "azurerm_storage_account" "avd_manager" {
│ 
╵
```

### Expected Behaviour

Storage should be deployed without error

### Actual Behaviour

1. If storage account is created for the first time, everything works as expected
2. If storage account is created, then destroyed and then again created with the same name, there is random error comming out.

image

The behaviour:

- Even if the error is showed up in TF output, the storage account is deployed correctly.
- The error is probably caused by DNS propagation / delay once the account is created. I assume the container client can not be created imidiately after the storage account is created due to the DNS propagation delay.
- This has been working fine like half year ago and i guess there has been PR to remove silently consuming those error (see link to the PR bellow)
- There is no way out of this error, except renaming the storage account resources (and all resources deppending on the account, i.e. function app)

### Steps to Reproduce

1. terraform apply 
2. terraform desttroy 
3. terraform apply 
repeat 1,2,3

### Important Factoids

_No response_

### References

Probably this is the PR which changed the provider / resource behaviour - https://github.com/hashicorp/terraform-provider-azurerm/pull/19062/files

### LINKED ISSUE #23240: Unable to refresh storage account state
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

v1.5.7

### AzureRM Provider Version

v3.60.0

### Affected Resource(s)/Data Source(s)

azurerm_storage_account

### Terraform Configuration Files

```hcl
resource "azurerm_storage_account" "mainstore" {
  name                     = "somethingst"
  resource_group_name      = azurerm_resource_group.rg.name
  location                 = azurerm_resource_group.rg.location
  account_tier             = "Standard"
  account_replication_type = "GRS"


  blob_properties {
          change_feed_enabled           = false
        #   change_feed_retention_in_days = 1
          last_access_time_enabled      = false
          versioning_enabled            = false
        }
  timeouts {
    read = "10m"
  }
}


resource "azurerm_role_assignment" "storage_ra1" {
  scope                = azurerm_storage_account.mainstore.id
  role_definition_name = "Storage Blob Data Contributor"
  principal_id         = azurerm_kubernetes_cluster.aks.identity[0].principal_id
  depends_on = [ 
        azurerm_kubernetes_cluster.aks
   ]
}
resource "azurerm_resource_group" "rg" {
  name     = var.global_prefix
  location = var.region
}
```

### Debug Output/Panic Output

```shell
terraform plan -refresh-only -target=azurerm_storage_account.mainstore -var-file=defaults.tfvars
azurerm_resource_group.rg: Refreshing state... [id=/subscriptions/0c4c0d8d-df06-4135-b9b7-a844966546ce/resourceGroups/tiportal]
azurerm_storage_account.mainstore: Refreshing state... [id=/subscriptions/0c4c0d8d-df06-4135-b9b7-a844966546ce/resourceGroups/tiportal/providers/Microsoft.Storage/storageAccounts/tiportalst]

#### INDEFINITELY HANGS HERE #####
#### CTRL-C                  #####
Stopping operation...

Interrupt received.
Please wait for Terraform to exit or data loss may occur.
Gracefully shutting down...


Planning failed. Terraform encountered an error while generating this plan.

╷
│ Warning: Resource targeting is in effect
│
│ You are creating a plan with the -target option, which means that the result of this plan may not represent all of
│ the changes requested by the current configuration.
│
│ The -target option is not for routine use, and is provided only for exceptional situations such as recovering from
│ errors or mistakes, or when Terraform specifically suggests to use it as part of an error message.
╵
╷
│ Error: reading queue properties for AzureRM Storage Account "tiportalst": queues.Client#GetServiceProperties: Failure sending request: StatusCode=0 -- Original Error: context canceled
│
│   with azurerm_storage_account.mainstore,
│   on storage.tf line 1, in resource "azurerm_storage_account" "mainstore":
│    1: resource "azurerm_storage_account" "mainstore" {
│
╵


Gist: https://gist.github.com/truekonrads/a6e7d35dc13b3f530cabf0a898c836d2
```

### Expected Behaviour

State is refreshed

### Actual Behaviour

Indefinitely hangs

### Steps to Reproduce

1. Create the resource (terraform apply)
2. Refresh the resource (terraform refresh or plan -refresh-only)

### Important Factoids

_No response_

### References

_No response_

### LINKED ISSUE #25160: `azurerm_storage_account` should display correct error message if it creation fails due to different Region of VNET Subnet ID
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

The creation of an Azure Storage Account can fail if it is created with a VNET Link to a resource in a different Azure region.

Example:
Storage Account creation in Germany West Central
VNet Link to resource in West Europe

The provider only prints a very generic:

image

The Storage Account can then be found in the 'Deleted' section. Especially annoying if you have enabled Purge Protection.

Enabling TF_DEBUG shows the real reason then:

`
"status":"Failed","error":{"code":"NetworkAclsValidationFailure","message":"Validation of network acls failure: ResourceBeingAcledHasWrongLocation:Microsoft.Storage resources in germanywestcentral cannot be ACL-ed to virtual network /subscriptions/xxx/resourceGroups/yyy/providers/Microsoft.Network/virtualNetworks/zzz-virtual-network in westeurope. Only resources in westeurope, northeurope can be ACL-ed to virtual networks in westeurope.."}}: timestamp=xxxx
`

The real error message should be displayed by the provider in this case.

### New or Affected Resource(s)/Data Source(s)

azurerm_storage_account

### Potential Terraform Configuration

```hcl
Important is only to set the following (referencing resource in a different region):


  network_rules {
    ...
    virtual_network_subnet_ids = var.virtual_network_subnet_ids
```
```


### References

_No response_

### LINKED ISSUE #26389: large_file_share_enabled parameter in storage_account
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

1.5.7

### AzureRM Provider Version

3.108.0

### Affected Resource(s)/Data Source(s)

azurerm_storage_account

### Terraform Configuration Files

```hcl
large_file_share_enabled = false
```

### Debug Output/Panic Output

```shell
no warning or error message
```

### Expected Behaviour

setting Enable large file shares option to false and 5TiB in azure portal

### Actual Behaviour

this option is forced in to be set to true and 100TiB by microsoft as per message in portal : 
"Provides file share support up to a maximum of 100 TiB. Large file shares are enabled by default on all storage accounts and cannot be disabled."

therefore the parameter itself is not useable anymore in the way how it's described in documentation.

### Steps to Reproduce

create storage account.

### Important Factoids

_No response_

### References

https://learn.microsoft.com/en-us/azure/storage/files/storage-how-to-create-file-share?tabs=azure-portal#advanced

### PR DESCRIPTION
**Pull request by tombuildsstuff**
This PR refactors the `azurerm_storage_account` Data Source and Resource and the `azurerm_storage_account_network_rules` Resource over to using `hashicorp/go-azure-sdk`.

This PR supersedes #23002 by @petr-stupka, since the base layer used in that PR is no longer used (and therefore the logic in it will no longer work) - and also supersedes #25789 by @wuxu92, since that’s only related to the Static Website endpoint. However this PR fixes #23240, #22992, #19055, #18897 and #15609 (which both were trying to fix in slightly different manners) by polling during the creation of the Storage Account, when applicable, for each type of supported sub-service (e.g. Blobs, Files, Static Websites etc) using a custom poller.

This also fixes #25160 and #12627 by switching to use `hashicorp/go-azure-sdk`’s base layer, which’ll correctly surface errors coming from the Azure API.

Fixes #25160
Fixes #23240
Fixes #22992
Fixes #22507
Fixes #19427
Fixes #19055
Fixes #18897
Fixes #15609
Fixes #14737
Fixes #12627
Fixes #26389

---

Changelog

```
(TBD)
```

### CODE REVIEW

**Comment by sharninder-se** [comment]
@tombuildsstuff Any chance I can get you to look at this PR ? I've been hitting this consistently while creating new storage accounts and I can't figure if its my setup or azure that is the problem.

File: internal/services/storage/storage_account_resource.go [hunk 1]
```diff
@@ -2077,288 +1959,255 @@ func resourceStorageAccountRead(d *pluginsdk.ResourceData, meta interface{}) err
...
-			d.Set("cross_tenant_replication_enabled", crossTenantReplication)
-		}
+		if props := model.Properties; props != nil {
+			primaryEndpoints = props.PrimaryEndpoints
+			routingPreference = props.RoutingPreference
+			secondaryEndpoints = props.SecondaryEndpoints
 
-		// There is a certain edge case that could result in the Azure API returning a null value for AllowBLobPublicAccess.
-		// Since the field is a pointer, this gets marshalled to a nil value instead of a boolean.
+			d.Set("access_tier", props.AccessTier)
```
Comment (manicminer) [inline internal/services/storage/storage_account_resource.go:2030]:
```suggestion
			d.Set("access_tier", pointer.From(props.AccessTier))
```

File: internal/services/storage/storage_account_resource.go [hunk 2]
```diff
@@ -2077,288 +1959,255 @@ func resourceStorageAccountRead(d *pluginsdk.ResourceData, meta interface{}) err
...
+			d.Set("allowed_copy_scope", props.AllowedCopyScope)
```
Comment (manicminer) [inline internal/services/storage/storage_account_resource.go:2031]:
```suggestion
			d.Set("allowed_copy_scope", pointer.From(props.AllowedCopyScope))
```

**Review by manicminer — APPROVED** [review]
Sorry for the delay in reviewing @tombuildsstuff. This LGTM, just a linter to appease and a question inline regarding ID parsing for MHSM vs Storage.

**Comment by krapgras** [comment]
Any update on the timeline for this fix?

**Review by manicminer — APPROVED** [review]
Have added some linting/naming changes and minor fixes. LGTM pending tests.

**Comment by manicminer** [comment]
Test results:

### CDN

Screenshot 2024-07-18 at 21 25 40

### Compute
_failures are 4.0 related_

Screenshot 2024-07-18 at 21 25 46

### Storage

Screenshot 2024-07-18 at 21 36 47

Screenshot 2024-07-18 at 21 37 47
<br><br>

_remaining failures are 4.0 related_

Screenshot 2024-07-18 at 21 38 15

**Pull request merged**